### PR TITLE
optimize kernel index

### PR DESCRIPTION
**Pull request by snnn**
**Description**: Describe your changes.
Currently session_kernels_ has all the kernels in an unordered_map and GetKernel does a lookup by nodeid.

As the nodeid to kernel mapping won't change we could instead build a std::vector<OpKernel*> with index of nodeid

**Motivation and Context**
- Why is this change required? What problem does it solve?

To make GetKernel run faster

- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnxruntime/core/framework/session_state.h [hunk 1]
```diff
@@ -187,15 +199,14 @@ class SessionState {
...
-  void CalculateNodeIndexInfo();
   const NodeIndexInfo& GetNodeIndexInfo() const;
 
  private:
   ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(SessionState);
 
   // cache of the constructed kernels to avoid spending construction
   // time per executor
-  std::unordered_map<NodeIndex, std::unique_ptr<OpKernel>> session_kernels_;
+  std::vector<OpKernel*> session_kernels_;
```
Comment (snnn) [inline onnxruntime/core/framework/session_state.h]: This line is the major change.

File: onnxruntime/core/framework/session_state.cc [hunk 2]
```diff
@@ -215,19 +285,8 @@ const SessionState* SessionState::GetSubgraphSessionState(onnxruntime::NodeIndex
...
-    for (auto& attr_name_to_subgraph : node_to_map_pair.second) {
-      attr_name_to_subgraph.second->CalculateNodeIndexInfo();
-    }
-  }
-}
-
 const NodeIndexInfo& SessionState::GetNodeIndexInfo() const {
-  ORT_ENFORCE(node_index_info_, "CalculateNodeIndexInfo must be called prior to GetExecutionInfo.");
+  ORT_ENFORCE(node_index_info_, "SetGraphAndCreateKernels must be called prior to GetExecutionInfo.");
   return *node_index_info_;
```
Comment (skottmckay) [inline onnxruntime/core/framework/session_state.cc]: This should be updated to 'prior to GetNodeIndexInfo'

File: onnxruntime/core/framework/session_state.h (hunk 1 above)
Comment (skottmckay) [reply onnxruntime/core/framework/session_state.h]: Why are we not using unique_ptr so the ownership semantics are clear?

---
In reply to: [316516508](https://github.com/microsoft/onnxruntime/pull/1672#discussion_r316516508) [](ancestors = 316516508)

**Comment by pranavsharma** [comment]
@snn I'm curious to know how much perf gain is noticed after this change? Also, I'm assuming there won't be any gaps in the node indices or in other others the vector will be densely populated, yes?

**Comment by snnn** [comment]
On resnet50, it's about 0.4%

**Comment by snnn** [comment]
It's not totally dense. But It won't waste too much memory. Because each element is just at most 8bytes. And, if it is really needed to save that, there is a way.

**Pull request merged**